django_mongodb/query.py

```diff
@@ -302,3 +305,124 @@ def register_nodes():
+            sql=raw_query, using=self.db, model=self.model, params=params
+        )
+
+    def iterator(self):
+        yield from MongoRawModelIterable(self)
+
+    def resolve_model_init_order(self, columns):
+        """Resolve the init field names and value positions."""
+        model_init_fields = [
+            f for f in self.model._meta.fields if f.column in columns
+        ]
+        annotation_fields = [
+            (column, pos)
+            for pos, column in enumerate(columns)
+            if column not in self.model_fields
+        ]
+        model_init_order = [
+            columns.index(f.column) for f in model_init_fields
+        ]
+        model_init_names = [f.attname for f in model_init_fields]
+        return model_init_names, model_init_order, annotation_fields
+
+
+class MongoRawModelIterable(RawModelIterable):
+    """
+    Iterable that yields a model instance for each row from a raw queryset.
+    """
+
+    def __iter__(self):
+        # Cache some things for performance reasons outside the loop.
+        db = self.queryset.db
+        query = self.queryset.query
+        connection = connections[db]
+        compiler = connection.ops.compiler("SQLCompiler")(query, connection, db)
+        query_iterator = self._make_result(query)
+        # FIXME Get columns from first row of the query without iterating over
+        # the query.
+        columns = self.queryset.columns
+
+        try:
+            (
+                model_init_names,
+                model_init_pos,
+                annotation_fields,
+            ) = self.queryset.resolve_model_init_order(columns)
+            model_cls = self.queryset.model
+            if model_cls._meta.pk.attname not in model_init_names:
+                raise FieldDoesNotExist(
+                    "Raw query must include the primary key"
+                )
+            fields = [self.queryset.model_fields.get(c) for c in self.queryset.columns]
+            converters = compiler.get_converters(
+                [f.get_col(f.model._meta.db_table) if f else None for f in fields]
+            )
+            if converters:
+                query_iterator = compiler.apply_converters(query_iterator, converters)
+            for values in query_iterator:
+                # Associate fields to values
+                model_init_values = [values[pos] for pos in model_init_pos]
+                instance = model_cls.from_db(db, model_init_names, model_init_values)
+                if annotation_fields:
+                    for column, pos in annotation_fields:
+                        setattr(instance, column, values[pos])
+                yield instance
+        finally:
+            # Done iterating the Query. If it has its own cursor, close it.
+            if hasattr(query, "cursor") and query.cursor:
+                query.cursor.close()
+
+    def _make_result(self, query):
+        for result in query:
+            yield list(result.values())
```

Comment:
How can we sure that the values are in the right position?

Reply:
the columns from the queryset columns could help to sort the returned values

Reply:
That happens in `resolve_model_init_order` IIUC but it fails because Django expects that `columns == fields` when the documents return from MongoDB don't always have all the fields. So we need a way to get the columns of the first row without iterating over the results of the Django query. What about an "escape hatch" query to get the first row directly from MongoDB as needed?

IOW

```
try:
 resolve_model_init_order(columns)
except:
  columns = escape_hatch(query)
  resolve_model_init_order(columns)

```


?

Reply:
I really think it ought to be fixable without an extra query. The code may need a logical rework so that the work `resolve_model_init_order()` does happens on the first iteration rather than before the loop starts.